### DISCUSSION TITLE
HTCondor tutorial

### DISCUSSION THREAD

**Discussion by carlkross** [GoogleCloudPlatform/cluster-toolkit#660] [Q&A] [answered]
I am wanting to work though the HTCondor totorial. There was a failure on my first attempt. I'm sorry I did not note it down.

Now I want to make a second attempt. I removed all that I could find regarding my first attempt but now I get several 409 errors saying that various features already exist.

I can provide more information if that is helpful but I wondered if there is a basic clean-up step that I am missing?

Thanks, Carl Ross

**Comment by tpdownes** [comment, marked as answer]
Hi Carl!

Thanks for trying the HTCondor tutorial. My guess, based on what you are telling me is that you tried the HTCondor tutorial using Cloud Shell (the command line tool in our web interface) and that you manually removed resources after provisioning a first go? Then you tried to provision a 2nd set?

1. Was the initial failure when you ran terraform or after SSH-ing to a machine?
1. Could you print the errors and the commands that produce the errors?
2. The "terraform way" of removing resources is to run `terraform destroy` when you are done with your work. Did you do that?
3. One way to avoid naming collisions in the toolkit is to use a different deployment_name.

You can do (3) by changing the deployment_name inside the blueprint YAML or by supplying an override at the command line.

e.g. `ghpc create --vars deployment_name=new-example --vars project_id=my-project file.yaml` 

We'd also love to know more about what your overall objective is.

**Reply by carlkross** [reply]
Thank you for your quick and detailed reply. You ask about my objective so perhaps I will start there.

I am a (retired) physicist working as a consultant for a company that requires Monte Carlo calculations (radiation transport). Such calculations are easily run in parallel, I have them running on a desktop (8 cores) and recently on a Google Cloud VM (112 cores).

I would like to consider the option of running a cluster with, maybe, 1000 cores, for at least one problem.

I know about containers and have had my application running in one. It then seemed like it should be easy to set up a cluster. However, after looking at Google Batch and then Kubernetes I came to realize that it is not so simple.

HTCondor seems hopeful and has the advantage that I can get it running on my desktop. However, working through the tutorial was quite daunting.

I was trying to stick within my single Google Cloud project, for which I already have quota allocated but maybe that was a mistake.

Since writing, I created a second project and was able to get through to the end of the tutorial, although the example still does not run. It just seems to stay in "idle" indefinitely.

Now, to your questions:

Yes, the failures were with terraform.

I should have saved the errors, but didn't. There were several 409 errors.

I didn't know about terraform destroy. That seems useful.

I did try a different  deployment name but it seemed like "htcondor-001" was hardcoded in elsewhere. But maybe that is because I didn't know about terraform destroy.

I feel I have made progress by creating a second project but I need to get the simple example to run.

Then, I need to learn how to select and expand my compute nodes. I expect that is done in the .yaml file but I still have a lot to learn.

Thanks,

Carl

**Reply by carlkross** [reply]
As a follow-up, to get the errors I saw, I restarted the tutorial.

I did "terraform destroy" but it said there was nothing left to destroy.

Here is a small file with all the 409 errors. They all seem related to things already existing (which maybe isn't really an error ...).

[Tutorial_errors.txt](https://github.com/GoogleCloudPlatform/hpc-toolkit/files/9873196/Tutorial_errors.txt)

**Comment by tpdownes** [comment]
You might find [designing a blueprint](https://cloud.google.com/hpc-toolkit/docs/setup/hpc-blueprint#design-blueprint) to be a useful link to look at. The idea is that `blueprint_name` describes a repeating pattern, while `deployment_name` is a specific instance of that pattern (e.g. "htcondor-pool" and "carl-test"). Using these examples, "carl-test" is a "deployment" of "htcondor-pool".

What I'm trying to show you above is that you can set the deployment_name in the YAML file or at the command-line. e.g.

```
ghpc create --vars deployment_name=carl-pool --vars project_id=my-project file.yaml
```

We have a couple paths forward:

1. Cleanup (destroy) the components that remain from your first attempt and try again "fresh"
2. Import the components that remain from your first attempt into a new deployment with the same name ("htcondor-001")

(1) is probably the simplest. (2) might teach you a bit about how terraform works if you're eager to learn.

There is also a 3rd path. I do think that we should discuss Batch would be more appropriate for you. An advantage of Batch is there may be no infrastructure for you to maintain. Are you already running your job containerized in HTCondor? Our Batch solution can create a "login node" in which you can test your job and then submit them at larger scale.

The magic of both Batch and our HTCondor solutions is that the compute nodes expand and contract based upon job submissions. So you don't need to say in advance how many nodes you want. They will be turned on to run a job and powered down when no jobs remain. 

Let me know about (1) or (2); we should definitely get you through a clean `terraform apply` / `terraform destroy` cycle to ensure you don't have lingering costs.

**Reply by carlkross** [reply]
Yes, I would like to accomplish (1). As I mentioned, I tried to run the tutorial in my existing project, and I would prefer not to delete the whole project because I have a VM instance there that I use. (With hindsight I should have just created a separate project).

When I first started reading about Batch I thought it was exactly what I needed. I could see how it could run several identical jobs (or maybe the correct terminology is tasks) on a single VM instance (say, wtih 112 cores) but I did not discover how it expands to more instances.

If it can, then I should go back and understand how to do that. There seemed to be considerable less overhead than with HTCondor.

So I have a containerized version of my job that runs on my desktop. I had done that in preparation for getting some cluster configuration working (first I loooked at Kubernetes, then Batch and finally HTCondor). But no, I have not run my container in HTCondor.

A general question regarding quota and instances. So far, I haven't requested more than 112 cores so that I can run on a single c2d machine. However, if I get to the point of wanting, say, 1000 or more, what is the logic guiding a machine type? Presumably I want c2d machines and the simplest would be to select 112 cores. But the same machine is available with fewer cores. Or are you just assigned a range of whatever c2d machines are available?

Thanks,

Carl

**Comment by tpdownes** [comment]
I think the following manual steps are shortest path to cleaning up. Browse to each URL and take the action noted. You will probably need to make sure that your project is selected the first time, but then it should "stick".

- https://console.cloud.google.com/security/secret-manager
  - delete htcondor-001-pool-password
- https://console.cloud.google.com/compute/disks
  - delete access-point-boot-disk-0
  - delete central-manager-boot-disk-0
- https://console.cloud.google.com/iam-admin/serviceaccounts
  - delete htcondor-001-access
  - delete htcondor-001-cm
  - delete htcondor-001-execute
- https://console.cloud.google.com/networking/addresses/list
  - delete htcondor-pool-nat-ips-us-central1-0
  - delete htcondor-pool-nat-ips-us-central1-1
- https://console.cloud.google.com/net-services/nat/list
  - delete anything matching htcondor
- https://console.cloud.google.com/hybrid/routers/list
  - delete anything matching htcondor
- https://console.cloud.google.com/networking/networks/list
  - delete htcondor-pool

Regarding Batch and VM type selection...

Are the tasks fully independent? e.g. if you run on a 112 core machine you are running 112/N jobs, where N is the number of threads per job. You are running the containerized version under Docker?

**Reply by carlkross** [reply]
Thanks for your help!!

Everything started off fine, even my project was selected.

However, when I got to addresses/list, I first had to delete the routers and then I could delete the addresses.

Then, when I get to networks/list I see (on two single lines)

htcondor-pool		1	8896	Custom	
None	
2	On		
us-central1	htcondor-pool-usc1			
10.0.0.0/24	
None	
None	10.0.0.1			Off

When I attmept to delete I get:

The deletion of the subnet failed. Error: The subnetwork resource 'projects/flkforflt/regions/us-central1/subnetworks/htcondor-pool-usc1' is already being used by 'projects/flkforflt/regions/us-central1/instanceGroups/htcondor-001-x-mig'

I went through the list again but I didn't see this condor reference. Does it refer to the compute pool that was created?

Sorry I got myself in such a mess.

Carl

**Reply by carlkross** [reply]
OK - looking in Instance Groups I found htcondor-001-x-mig.

I deleted it and then could delete the networks.

Maybe now it's all gone?

**Comment by tpdownes** [comment]
I'd like to begin by suggesting you run a non-containerized job using Batch:

- https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/docs/cloud-batch.md
- https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/examples/README.md#cloud-batchyaml--
- https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/examples/cloud-batch.yaml

The blueprint above provisions:

- a shared filesystem
- a VM that acts like the Cloud Batch environment and mounts the shared filesystem at `/sw`
- produces a job template that reads in a file from the shared filesystem

You could remove the shared filesystem if it's not useful for you, but it might be a useful place to install non-containerized code from the VM environment.  Also a useful place for output data. We can then discuss a containerized approach if you'd like.

**Reply by carlkross** [reply]
You asked about my tasks. Taking the example of my desktop (8 cores) my submissions look like:

my_exec some switches input_file_001.imp &
my_exec some switches input_file_002.imp &
on up to 008.

The input files all describe the same problem; they differ only in the random number seed so that all eight runs are statistically independent. (I have the same running on a VM instance of c2d with 112 cores).

These all run independently and complete in more or less the same run time.

Each run outputs a series of files that are combined statistically.

I will look again at Batch. Clearly, I need some sort of file system to hold the input files and the output files.

Are you suggesting that I can somehow just use my existing c2d image and run it in Batch mode? I had assumed I needed my executable in a container.

Am I right that Batch is limited to 1000 cores? This is likely fine of my application but I guess HTCondor is unlimited?

Thanks, Carl

**Reply by nick-stroud** [reply]
Hey @carlkross, 

**file system**: 
The [batch example](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/examples/cloud-batch.yaml) referenced by @tpdownes contains a [filestore](https://cloud.google.com/filestore#documentation) file system that will be mounted to each Batch job. In the example it is mounted under `/sw` but the specific local mount is [configurable](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/06de956de90fb88cb44bc30026cc144e3dbe3a39/community/examples/cloud-batch.yaml#L36). The login node created by that blueprint will also have access to the filestore. You can use the login node to stage data to the shared filestore that is needed by the batch jobs and inspect outputs written to the shared filestore.

**Image**:
If you have a VM image (as opposed to a docker image) then you can simply reference it in [these lines](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/06de956de90fb88cb44bc30026cc144e3dbe3a39/community/examples/cloud-batch.yaml#L57-L58) and your batch job will use that image. Your blueprint would include something like:
```
  - id: batch-job
    source: community/modules/scheduler/cloud-batch-job
    use: [network1, appfs]
    settings:
      runnable: <<Execute your job>>
      machine_type: c2d-standard-4  # updated to c2d machine type
      image:
        family: <<Your VM Image>>
        project: <<Your Project ID>>
```

This might not work if your job is reliant on running inside a container, or if your image is not a VM image but instead a Docker image. 

**Cores**:
Batch restrictions are documented [here](https://cloud.google.com/batch/quotas#job_limits). It looks like Batch is restricted to 1000 parallel tasks. Each task could be running on a machine with multiple cores, so I believe the actual core limitation would be much higher.

**Reply by carlkross** [reply]
Thanks for your help, Nick. I have my code running on a c2d machine so I should be able to use that VM image.

The terminology of jobs, task and runnables is a bit confusing to me. Before I start, can you help me with the entry 

runnable: <<Execute your job>>

I attach here an example script (with generic names) of what I run on my 8 core desktop. I can do the same on a c2d-112 machine where now I have 112 lines.

In my case, is this script my runnable?

Suppose it is and now I have 225 entries in the script. Assume I have been allocated at least 225 cores. Does Batch automatically arrange extra cores so all my jobs run in parallel?

(The limit seems clear: 1000 parallel tasks. I don't understand a task group. Anyway, we can leave that for another time).

[example.txt](https://github.com/GoogleCloudPlatform/hpc-toolkit/files/9907082/example.txt)

In real life, this is a .sh file but it said I was not allowed to paste that file type.

**Reply by nick-stroud** [reply]
I believe you have two options:

**Option 1: Single line runnable: one task per core:**
In this option we put the actual job execution (`/path-to/my_exec`) as the runnable. 

General steps:
1. Deploy the blueprint
2. SSH into the login node and stage your data to `/share`. Note: you may have to wait a few minutes after deployment before `/share` is available on the login node. 
3. Submit the Batch job

Things to note:
- This is using the `$BATCH_TASK_INDEX` to reference the file name. These are indexed from 0.
- This assumes my_exec is built into the image. Alternatively you could put in /share along with the data.
- The output data needs to be written/copied to `/share` or it will be lost when the job finishes.
- This works well until we want to use more than 1000 cores and we run into the task limit.
- You asked about 225 cores but multiples of the cpu count would work best so I used 224.

```
  modules:
  - id: network1
    source: modules/network/pre-existing-vpc

  - id: sharedfs
    source: modules/file-system/filestore
    use: [network1]
    settings: {local_mount: /share}

  - id: batch-job
    source: community/modules/scheduler/cloud-batch-job
    use: [network1, sharedfs]
    settings:
      runnable: "/path-to/my_exec -M 5 /share/input_file_`printf %04d $BATCH_TASK_INDEX`.inp > /dev/null"
      machine_type: c2d-standard-112
      task_count: 224 
      task_count_per_node: 112
      image:
        family: <<Your VM Image>>
        project: <<Your Project ID>>

  - id: batch-login
    source: community/modules/scheduler/cloud-batch-login-node
    use: [batch-job]
    outputs: [instructions]
```

**Option 2: Runnable calls a script: one task per machine:**
In this option the runnable calls your script, which would then call `my_exec` on 112 input files. 

Things to note:
- In this option you have the additional burden of writing `execution_script_0.sh` and `execution_script_1.sh` which would each call `my_exec` on a different set of input files.
- The task count is an order of magnitude smaller since you only have 1 task per machine instead of 1 task per core. This would allow you to scale to much larger jobs. 

```
  modules:
  - id: network1
    source: modules/network/pre-existing-vpc

  - id: sharedfs
    source: modules/file-system/filestore
    use: [network1]
    settings: {local_mount: /share}

  - id: batch-job
    source: community/modules/scheduler/cloud-batch-job
    use: [network1, sharedfs]
    settings:
      runnable: "/share/execution_script_${BATCH_TASK_INDEX}.sh"  # similar to your example
      machine_type: c2d-standard-112
      task_count: 2
      task_count_per_node: 1
      image:
        family: <<Your VM Image>>
        project: <<Your Project ID>>

  - id: batch-login
    source: community/modules/scheduler/cloud-batch-login-node
    use: [batch-job]
    outputs: [instructions]
```

In both cases you should be able to test the runnable on the login node (you will have to set the `BATCH_TASK_INDEX` yourself). Note that I have not actually deployed and tested these blueprints so report back if you hit issues.

**Comment by tpdownes** [comment]
@carlkross we are going to close this discussion, but please do not hesitate to respond if you would like further assistance, esp. with Batch solution. My recommendation would be to try the example without (for now) considering your application and submit a few "hello, world!" jobs.

**Reply by carlkross** [reply]
Thanks for your help. Based on Nick's suggestion, I have looked at Batch again. I am not convinced that it is the right solution for me. Perhaps it is the terminology (jobs, tasks, runnables, groups of tasks) that I don't properly understand.

So I have returned to Condor. I now have it running with my container on an eight-core desktop and it behaves as expected. Of course, in this case, the master node and execute nodes are the same machine but it bodes well for a cluster.

There is a lot to understand about deploying a cluster on Google Cloud but I should be able to do it, using the examples for help.

Things I am still struggling with:

- Do all the execute nodes have to be the same? E.g, c2-standard-4? If so, what is the "best " choice? (I'm assuming here that quota is not an issue).

- In the example htcondor.yaml, I see a choice for the central manager machine, but then a choice for an "access point" machine. Is this the same as an execute machine?

- There is a "content" entry that looks a lot like a Condor "submit" file. Does this have to be part of the .yaml file?

I will continue to look at the documenation and examples.

Carl

**Reply by tpdownes** [reply]
First, the HTCondor team has updated some of their terminology for better/worse. Roughly:

```
"access point" = "login node" where you submit jobs
"central manager" = "master node" that schedules jobs (there is a daemon called condor_master which runs on every node and is entirely different)
"execute point" = "compute node" where jobs run
```

1. Presently the HTCondor solution in the Toolkit only supports a pool of homogenous "execute points". For CPU-only applications, we recommend the following for best performance because they occupy full NUMA nodes in the underlying hardware ([noisy neighbor](https://www.techtarget.com/searchcloudcomputing/definition/noisy-neighbor-cloud-computing-performance)). It sounds like the scale of your workload could occupy these nodes fully.
  - Intel: c2-standard-30, c2-standard-60
  - AMD: c2d-standard-56, c2d-standard-112 (any time you see `d` it means AMD)

2. Hopefully the terminology above addressed the issue. At your scale of nodes, you can pick fairly small access point and central manager nodes (n2-standard-4 or n2d-standard-4 ought to be fine)
3. There is no need to use a Toolkit `data` runner to pre-stage a submit file. It's nice (from our perspective) for the tutorial or perhaps from yours if you are sharing your solution with other people.

**Reply by carlkross** [reply]
Thanks for your help and sorry for the long delays in responding and trying things. I was travelling and also had other duties.

I am now trying to develop some understanding of the .yaml  blueprint file. I can see where the access point machine is defined:

settings:
      name_prefix: access-point
      machine_type: c2-standard-4

and the central manager machine:

settings:
      name_prefix: central-manager
      machine_type: c2-standard-4

Although I see references to condor execute points I have yet to discover where I define the execute machines.

My initial goal is to see that I can set up a very basic cluster with an n2-standard-2 as an access machine, the same for a central manager and a c2-standard-4 as a single compute node. I expect that if I can get that going then I am off to the races.

**Reply by cboneti** [reply]
Hi Carl, 

Tom and Nick are OOO this week. I will try to repro your deployment tomorrow and come back to you.

Thanks!

Carlos

**Reply by nick-stroud** [reply]
Hey Carl,

I assume you are looking at the [htcondor-pool.yaml](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/examples/htcondor-pool.yaml) example. 


**Central Manager and Access Point**  
It seems like you have found where to set the `machine_type` for the central manager and access point (line [62](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/examples/htcondor-pool.yaml#L62) & [120](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/examples/htcondor-pool.yaml#L120)). To implement your idea for a basic cluster you would need to update these lines to use `n2-standard-2` instead of their current values of `c2-standard-4`.

**Execute Points**  
The execute points are created by [this block](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/examples/htcondor-pool.yaml#L78-L90) using the `htcondor-execute-point` module. The inputs for this module are defined [here](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/modules/compute/htcondor-execute-point/README.md#inputs). You can see that this module also has a `machine_type` input which defaults to `n2-standard-4`. Since this value is not set in the blueprint it will pick the default value. If you would like to specify the `machine_type` for the execute points then you would update the blueprint around line [83](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/examples/htcondor-pool.yaml#L83) to add the `machine_type` setting. See example code block below. 

```yaml
  - id: htcondor_execute_point
    source: community/modules/compute/htcondor-execute-point
    use:
    - network1
    - htcondor_configure_execute_point
    settings:
      machine_type: c2-standard-4  ## this line was added ##
      max_size: 1                  ## this line was added ##
      metadata:
        enable-oslogin: "TRUE"
        central-manager: ((module.htcondor_cm.internal_ip[0]))
      service_account:
        email: $(htcondor_configure.execute_point_service_account)
        scopes:
        - cloud-platform
```

The execute nodes are auto-scaling meaning that they won't actually be created until you submit a job. By default the `htcondor-execute-point` module will scale the node pool to 100. I have also added a setting above to limit the pool size to 1. 

Just FYI: In case you are not aware, there is an [HTCondor Tutorial](https://github.com/GoogleCloudPlatform/hpc-toolkit/tree/main/docs/tutorials#htcondor-tutorial) that uses the [htcondor-pool.yaml](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/examples/htcondor-pool.yaml) blueprint and has a step by step walkthrough.

**Reply by carlkross** [reply]
Hi Nick:

Thanks so much for helping guide me through this. Yes, I know about the tutorials and realize I have to dig deeper to better understand the yaml files.

First the good news - I see my access point and central manager created.
But I still have a couple of challenges.

Early on, I get a message that I have an IAM problem. I go to create credentials but I am not sure what catergory I am in. Do I have to create a new service?

The tutorial progressed anyway, saying it was going to treat the message as a warning.

When I try to SSH to the access point, I get

Connection via Cloud Identity-Aware Proxy Failed

It says I maybe can try anyway but then I get a network error:

Check firewall settings
No ingress firewall rule allowing SSH found.

Any guidance appreciated.

Carl

**Reply by carlkross** [reply]
Update:

I have succeded with an SSH connection to the access point. It seems that one may need to do a "retry" and then it connects.

I succeeded in running a simple non-container job with Condor. (It takes quite a while for the compute node to spin up, perhaps 5 minutes or more, so intially the situation is a bit confusing).

I succeeded in running the Hello-World container using Condor.

When it comes to my own container, it tells me that the container is not present locally. However, it is, and I can run it directly using Docker.

This remains an unresolved mystery at the moment.

Carl

**Reply by carlkross** [reply]
Update #2:

In thinking about it, it is not enough to have the container on the access point. It must be available to the compute nodes.

I created a Google container registry and pushed my image there. That worked and I can see my container image.

When I go to my Google Shell machine, I can pull the image.

However, when I go to the access point machine of my cluster and try to pull the image, I get

Error response from daemon: unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication

The documentation seems to state that if the cluster is part of the same project as the container registry (which it is) I should not have to do anything. Beyond that, the options are a bit overwhelming.

As I mentioned before, when building the cluster I got an IAM error (should have recorded it) but the build went on to treat it as a warning.

Should I make a fresh start and pay more attention to the IAM issue?

Carl

**Reply by nick-stroud** [reply]
Hey Carl,

**On boot up time:**
On boot up of the execute point, the machine is running a startup script ([blueprint section](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/examples/htcondor-pool.yaml#L71-L76)) which is installing HTCondor ([ansible playbook](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/0f0f70ea4282667ebae36a79558f0f8ce29ceb23/community/modules/scripts/htcondor-install/files/install-htcondor.yaml)) and configuring the execution point ([ansible playbook](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/main/community/modules/scheduler/htcondor-configure/files/htcondor_configure.yml)). Since this startup script is actually running ansible playbooks, it also has to install ansible which takes several minutes. 

My suspicion is that all of the above is why you are seeing 5+ min of boot up time. The startup time can be decreased by building an image that already has the dependencies installed (ansible, HTCondor, etc). There is support in the HPC Toolkit for image building and this may be a good option to explore in the future if your usage scales significantly, but is probably more trouble that it is worth while you are still testing out the platform. So, my recommendation would be to use as is for now and live with the longer boot time.

**On Container Registry Authentication:**
It looks like `Storage Object Viewer` permission is needed to pull the container. The HTCondor blueprint will create service accounts for the access point, central manager, and execute point which should all automatically be given the `Storage Object Viewer` permission by default. The service account for the execute point should look something like `htcondor-001-execute@<project_id>.iam.gserviceaccount.com`. I am wondering if your error on deployment could have something to do with these service accounts not getting the correct permissions. Can you go into `IAM` > `PERMISSIONS` and confirm that the `htcondor-001-execute` has the `Storage Object Viewer` role? If it is missing, then try adding it.

**Reply by carlkross** [reply]
Thanks, Nick. I'm OK with the 5 minutes for now or maybe forever. Initally, it led me to believe things were not working.

I've checked Storage Object Viewer and all looks good. I attach what I see for the storage bucket.

And, as I pointed out, I can pull the image using my Cloud Shell machine but not from the access point machine. There is some permission/access difference between Cloud Shell and the cluster.

The fact that the Hello-World container works is also interesting. I think it is on the public Docker hub so I'm guessing my container might work if I put it there. In the long run, I don't want it public.

If you have no other ideas, I will make a fresh build, this time noting all errors.

Carl

![Storage_bucket](https://user-images.githubusercontent.com/33606318/206063480-b756327b-28c0-489b-9efd-7e48f05f42cf.png)

**Reply by carlkross** [reply]
Status report:

So I deleted my first cluster attempt and built a second, keeping track of all the errors/warnings.

This is the first error I get:

iam.googleapis.com: service is disabled in project fltclstr2

This time, I enabled the API as per instructions. It says you may need credentials but it seems not in my case??

Then I proceeded to Terraform. The instructions says you may get errors and to keep doing "Apply". I got three different sets of errors but after three "Applies" success.

The cluster came up and I was able to run the same two tests that worked before. Then I put my container in a container registry.

However, as before, I get this error:

Hold reason: Error from slot1_1@htcondor-001-x-17mh.us-central1-b.c.fltclstr2.internal: Unable to find image 'gcr.io/fltclstr2/fluka:v2' locally Error response from daemon: unauthorized: You don't have the needed permissions to perform this operation, and you may have invalid credentials. To authenticate your request, follow the steps in: https://cloud.google.com/container-registry/docs/advanced-authentication

I tried pulling the image to the access point but again failure. In checking the Docker group, Condor was there but no user. I added me as a user and **then I could pull the image.**

Finally, I put my container on the public Docker hub and **then I had success.**

So ... there is some issue with Docker getting rights to access the container registry from the execute nodes.

This should be easy (??) to fix but at this point I don't know how.

Carl

**Reply by nick-stroud** [reply]
Hey Carl, I just want to acknowledge that I have seen your replies. I personally don't have a lot of experience with containers so I am working on getting support from others that do.

**Reply by nick-stroud** [reply]
Hey Carl, 

Apologies for the slow response. I was able to reproduce your problem and have a workaround/solution. 

I was able to add a startup script to the execute points which perform docker authentication with Google Container Registry and then pull a private container. This container is then available locally on the machine when the job runs. Since this script will run every time a node is booted up, it should be able to always pull the latest container. 

I have not yet figured out how to apply the permissions to the condor user, so for now only containers specified in the startup script will be available (the startup script runs as root). I think it should be possible to also give the condor user authorization, although that will take some more work/testing.

### The Solution

I am using the [standalone helper](https://cloud.google.com/container-registry/docs/advanced-authentication#standalone-helper). The other methods may also work but this is what I got working. 

The module below would replace the `htcondor_configure_execute_point` module defined on [L71 of htcondor-pool.yaml](https://github.com/GoogleCloudPlatform/hpc-toolkit/blob/3c03c9aa543a13759e0dcd764f27abee8669cd31/community/examples/htcondor-pool.yaml#L71). The `docker_auth_and_pull.sh` is the added script. 

```yaml
  - id: htcondor_configure_execute_point
    source: modules/scripts/startup-script
    settings:
      runners:
      - $(htcondor_install.install_htcondor_runner)
      - $(htcondor_configure.execute_point_runner)
      - type: shell
        destination: docker_auth_and_pull.sh
        content: |
          VERSION=2.1.5
          OS=linux
          ARCH=amd64
          curl -fsSL "https://github.com/GoogleCloudPlatform/docker-credential-gcr/releases/download/v${VERSION}/docker-credential-gcr_${OS}_${ARCH}-${VERSION}.tar.gz" \
          | tar xz docker-credential-gcr \
          && chmod +x docker-credential-gcr && sudo mv docker-credential-gcr /usr/bin/
          docker-credential-gcr configure-docker
          gcloud auth application-default login --quiet

          # List the private docker images you want to use
          #   Note you must update the project_id
          docker pull gcr.io/<project_id>/busybox 
```

After adding this I am able to run jobs with: `docker_image   = gcr.io/<project_id>/busybox`.

One more note: at one point during my debugging I believe I also had to add `gcloud auth application-default login --quiet` to the script, but in the final form it ran without this line, so I think it was unneeded and has been excluded from the snippet above.

**Reply by carlkross** [reply]
Thanks for your efforts, Nick. It looks like this should work and I will try it with my container.

I guess docker has been installed and configured in the the Centos 7 OS? One might guess that that configuration needs to be tweaked.

My progress continues -slowly. I wanted to set up a remote desktop to the access point to have easier access to tools. I have done this with other VMs using Chrome remote desktop. However, it turns out Centos doesn't support Chrome remote desktop. I then had to learn how to add firewall rules (I couldn't even ping to the access point). Anyway, I finally have access using Remmina on my Ubuntu desktop but it still needs some fine tuning.

Happy Holidays! Carl

**Reply by cboneti** [reply]
Hi Carl, 

Thanks. It sounds like you got it all working. 

Providing a remote desktop module is in our radar, thanks for the feedback here.

**Reply by carlkross** [reply]
Status report:

I have spent some time working with Nick's fudge regarding the Docker container registry. At first, it seemed to work fine. That was with a single compute node.

I then moved to testing with up to five c2d-standard-8 machines. Now, some compute nodes load the container and others give me the usual permissions error.

If I put my container on the Docker public hub all five machines work fine.

It is tedious to do testing and I'm not sure what to change or to look for.

BTW, I have left out this line in the script

gcloud auth application-default login --quiet

which Nick said was not required.

Ideally, I would like to keep my container private but at the moment that doesn't seem possible.

Carl
